### PR TITLE
Multiple signatures for transactions

### LINKED ISSUE #1531: Can't bundle AfterSignature and ApplySignature instructions in single transaction
Problem:
Can't currently implement multisig in a single transaction.

Proposed solution:
Change Transaction's `signature: Signature` to `signatures: Vec<Signature>`. Sigverify should use the first `signatures.len()` of `account_keys` for the respective pubkeys.

TODO: Decide if the signed data should be the same for all signatures, or if all signatures after the first simply sign the first signature. I don't have a strong opinion, but signing the first signature is what Budget's ApplySignature currently expected.

TODO: Does that rule about the first `signatures.len()` keys being signed create any problems? Does it alleviate the need for an instruction to specify `signed_account_keys`?

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

With multiple instructions in a TX may need
multiple signatures.

#### Summary of Changes

Add vec of signatures to a TX.

Fixes #1531

### CODE REVIEW

File: src/transaction.rs [hunk 1]
```diff
@@ -301,19 +315,22 @@ mod tests {
...
-                201, 177, 11, 117, 135, 187, 167, 181, 188, 22, 59, 206, 105, 231, 150, 215, 30,
-                78, 212, 76, 16, 252, 180, 72, 134, 137, 247, 161, 68, 32, 253, 186, 201, 177, 11,
-                117, 135, 187, 167, 181, 188, 22, 59, 206, 105, 231, 150, 215, 30, 78, 212, 76, 16,
-                252, 180, 72, 134, 137, 247, 161, 68, 1, 1, 1, 4, 5, 6, 7, 8, 9, 9, 9, 9, 9, 9, 9,
-                9, 9, 9, 9, 9, 9, 9, 9, 9, 8, 7, 6, 5, 4, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 99, 0, 0, 0, 0, 0,
-                0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 4, 5, 6, 7, 8, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-                1, 1, 1, 1, 1, 9, 8, 7, 6, 5, 4, 2, 2, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
-                0, 0, 0, 0, 1, 2, 3, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3
+                1, 0, 0, 0, 0, 0, 0, 0, 199, 111, 28, 46, 13, 235, 122, 203, 210, 118, 158, 0, 59,
```
[the remark this answers is not in the document]
Comment (mvines) [reply src/transaction.rs]: I can do it once this lands.  This test is just a signal that SDK changes will be required.  In this case there'll be a number of SDK API changes to expose the multiple signatures.  (cc: #1823)

File: src/transaction.rs [hunk 2]
```diff
@@ -38,8 +36,11 @@ impl Instruction {
 /// An atomic transaction
 #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
 pub struct Transaction {
-    /// A digital signature of `account_keys`, `program_ids`, `last_id`, `fee` and `instructions`, signed by `Pubkey`.
-    pub signature: Signature,
+    /// A set of digital signature of `account_keys`, `program_ids`, `last_id`, `fee` and `instructions`, signed by `signed_keys`.
+    pub signatures: Vec<Signature>,
+
+    /// The `Pubkeys` that correspond to signature in `signatures`
+    pub signed_keys: Vec<Pubkey>,
```
Comment (mvines) [inline src/transaction.rs]: The key duplication between `signed_keys` with `account_keys` is not great.  Seems like there'll be at least 1 duplicated key in almost every transaction.

Comment (garious) [reply src/transaction.rs]: I was thinking the first `signatures.len()` of `account_keys` would be the pubkeys used for signature verification.

**Comment by mvines** [comment]
Prior to this PR, all programs can (and generally do) assume that `account_key[0]` signed the transaction, which is (sometimes implicitly) used to enforce constraints like preventing the system program from moving funds from a 3rd party account.   With this PR, it seems like I could sign a Transaction containing a `system_move()` instruction that moves all your tokens to my account and bank would be good with it.

File: src/transaction.rs (hunk 2 above)
Comment (sakridge) [reply src/transaction.rs]: Well we can do that, when talking with @aeyakovenko I got the impression that we needed a separate `signed_keys` member and that you had proposed that.

Comment (garious) [reply src/transaction.rs]: And that `Transaction::signed_key()` would be changed from just key 0 to any index under `signatures.len()`.

Comment (garious) [reply src/transaction.rs]: I don't have record of it (we should have created an RFC), but I'd guess what was proposed was a separate `signed_keys` (named `signed_account_keys`) and that those same keys should *not* also appear in `account_keys`.  That'd be more intuitive but still a waste of 8 bytes.  I don't see a problem with both signed and unsigned pubkeys being in the same list and the length of the signatures being the trick to know which are which. No bit left behind.

Comment (sakridge) [reply src/transaction.rs]: @garious for `signed_key` is that an extra index for the signed key then:

```
pub fn signed_key(&self, instruction_index: usize, accounts_index: usize,
signature_index: usize
) -> Option<&Pubkey> {
    match self.key_index(instruction_index, accounts_index) {
        None => None,
        Some(0) => {
               if (signature_index < self.signatures.len() {
                    return self.account_keys.get(0);
               } else {
                    return None;
                }
        Some(_) => None,
    }
}
```

Comment (garious) [reply src/transaction.rs]: yep!

**Comment by garious** [comment]
@mvines, yes, those programs need to be upgraded to use `signed_key()` and `key()`. As an added benefit, that change will allow users to batch `Move` instructions as well.

**Comment by garious** [comment]
@sakridge, if you wanted to avoid opening that security hole before those programs were updated, you could reject transactions with `signatures.len() == 0` here: https://github.com/solana-labs/solana/blob/master/src/banking_stage.rs#L127

**Comment by sakridge** [comment]
@garious I think I would do it earlier in sig verify stage where it will be cheaper to reject. Is that the only condition needed?

**Comment by garious** [comment]
@sakridge, better idea, add a condition just before this one to check that key 0 is signed. Then it's not just a stopgap solution - it'll be permanent. The reason key 0 must be signed is that it's the one that pays the fee.  If downstream programs depend on that assumption, that won't be a problem.

**Comment by sakridge** [comment]
I think this is ready, will wait for @mvines sign off though.

**Pull request merged**